File: typescript/src/stream.ts
// MemoSiftStream — real-time compression stream, process messages as they arrive.

import {
  Pressure,
  pressure as computePressure,
  contextWindowFromModel,
  estimateTokensHeuristic,
} from "./core/context-window.js";
import type { AnchorFact } from "./core/types.js";
import type { CompressionReport } from "./report.js";
import { MemoSiftSession } from "./session.js";
import type { MemoSiftSessionOptions } from "./session.js";

/** Result of pushing a message to the stream. */
export interface StreamEvent {
  /** Either "buffered" (no compression triggered) or "compressed". */
  readonly action: "buffered" | "compressed";
  /** Whether compression was triggered on this push. */
  readonly compressed: boolean;
  /** Tokens saved by compression (0 if not compressed). */
  readonly tokensSaved: number;
  /** The current context pressure level. */
  readonly pressure: Pressure;
}

/**
 * Real-time compression stream — process messages as they arrive.
 *
 * Wraps a `MemoSiftSession` with incremental mode enabled.
 * Messages are buffered until context pressure warrants compression,
 * at which point compression runs only on new messages (Zone 3).
 *
 * ```typescript
 * const stream = new MemoSiftStream("coding", { model: "claude-sonnet-4-6" });
 *
 * for (const message of incomingMessages) {
 *   const event = await stream.push(message);
 *   if (event.compressed) console.log(`Saved ${event.tokensSaved} tokens`);
 * }
 *
 * const compressed = stream.messages;
 * const facts = stream.facts;
 * ```
 */
export class MemoSiftStream {
  private readonly _session: MemoSiftSession;
  private _messages: unknown[] = [];

  constructor(preset = "general", options?: Omit<MemoSiftSessionOptions, "incremental">) {
    this._session = new MemoSiftSession(preset, {
      ...options,
      incremental: true,
    });
  }

  /**
   * Push a new message and get a compression decision.
   *
   * The message is appended to the internal buffer. Compression is triggered
   * only when context pressure exceeds NONE.
   */
  async push(message: unknown): Promise<StreamEvent> {
    this._messages.push(message);

    const pressure = this._session.checkPressure(this.estimateTokens());
    if (pressure === Pressure.NONE) {
      return { action: "buffered", compressed: false, tokensSaved: 0, pressure };
    }

    const { messages, report } = await this._session.compress(this._messages);
    this._messages = [...messages];
    return {
      action: "compressed",
      compressed: true,
      tokensSaved: report.tokensSaved,
      pressure,
    };
  }

  /**
   * Force compression regardless of pressure.
   *
   * Useful at the end of a conversation or before a long pause.
   */
  async flush(): Promise<StreamEvent> {
    if (this._messages.length === 0) {
      return { action: "buffered", compressed: false, tokensSaved: 0, pressure: Pressure.NONE };
    }

    const { messages, report } = await this._session.compress(this._messages);
    this._messages = [...messages];
    const pressure = this._session.checkPressure(this.estimateTokens());
    return {
      action: "compressed",
      compressed: true,
      tokensSaved: report.tokensSaved,
      pressure,
    };
  }

  /** Current message state (may include compressed messages). */
  get messages(): readonly unknown[] {
    return [...this._messages];
  }

  /** Accumulated anchor facts. */
  get facts(): readonly AnchorFact[] {
    return this._session.facts;
  }

  /** The underlying session (for advanced configuration). */
  get session(): MemoSiftSession {
    return this._session;
  }

  /** Number of messages in the buffer. */
  get messageCount(): number {
    return this._messages.length;
  }

  private estimateTokens(): number {
    const contents: string[] = [];
    for (const msg of this._messages) {
      if (typeof msg === "object" && msg !== null && "content" in msg) {
        const c = (msg as Record<string, unknown>).content;
        contents.push(typeof c === "string" ? c : JSON.stringify(c ?? ""));
      } else {
        contents.push(String(msg));
      }
    }
    return estimateTokensHeuristic(contents);
  }
}

File: python/src/memosift/stream.py
# MemoSiftStream — real-time compression stream, process messages as they arrive.
from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Any

from memosift.core.context_window import Pressure, estimate_tokens_heuristic
from memosift.session import MemoSiftSession

if TYPE_CHECKING:
    from memosift.core.types import AnchorFact


@dataclass(frozen=True)
class StreamEvent:
    """Result of pushing a message to the stream.

    Attributes:
        action: Either "buffered" (no compression triggered) or "compressed".
        compressed: Whether compression was triggered on this push.
        tokens_saved: Tokens saved by compression (0 if not compressed).
        pressure: The current context pressure level.
    """

    action: str
    compressed: bool
    tokens_saved: int = 0
    pressure: Pressure = Pressure.NONE


class MemoSiftStream:
    """Real-time compression stream — process messages as they arrive.

    Wraps a ``MemoSiftSession`` with incremental mode enabled.
    Messages are buffered until context pressure warrants compression,
    at which point compression runs only on new messages (Zone 3).

    Usage::

        stream = MemoSiftStream("coding", model="claude-sonnet-4-6")

        for message in incoming_messages:
            event = await stream.push(message)
            if event.compressed:
                print(f"Saved {event.tokens_saved} tokens")

        # Get current compressed state
        compressed = stream.messages
        facts = stream.facts
    """

    def __init__(
        self,
        preset: str = "general",
        *,
        model: str | None = None,
        llm: Any = None,
        framework: str | None = None,
        **config_overrides: Any,
    ) -> None:
        """Create a new compression stream.

        Args:
            preset: Config preset name.
            model: Model name for context window lookup + adaptive compression.
            llm: Optional ``MemoSiftLLMProvider``.
            framework: Explicit framework. Auto-detected if omitted.
            **config_overrides: Config field overrides.
        """
        self._session = MemoSiftSession(
            preset,
            model=model,
            llm=llm,
            framework=framework,
            incremental=True,
            **config_overrides,
        )
        self._messages: list[Any] = []

    async def push(self, message: Any) -> StreamEvent:
        """Push a new message and get a compression decision.

        The message is appended to the internal buffer. Compression is triggered
        only when context pressure exceeds NONE (i.e., the model's context window
        is under meaningful load).

        Args:
            message: A framework-native message (dict, object, etc.).

        Returns:
            A ``StreamEvent`` indicating whether compression was triggered.
        """
        self._messages.append(message)

        # Check if compression is needed based on pressure.
        pressure = self._session.check_pressure(
            usage_tokens=self._estimate_tokens()
        )
        if pressure == Pressure.NONE:
            return StreamEvent(action="buffered", compressed=False, pressure=pressure)

        # Compress.
        compressed, report = await self._session.compress(self._messages)
        self._messages = list(compressed)
        return StreamEvent(
            action="compressed",
            compressed=True,
            tokens_saved=report.tokens_saved,
            pressure=pressure,
        )

    async def flush(self) -> StreamEvent:
        """Force compression regardless of pressure.

        Useful at the end of a conversation or before a long pause.

        Returns:
            A ``StreamEvent`` with compression results.
        """
        if not self._messages:
            return StreamEvent(action="buffered", compressed=False)

        compressed, report = await self._session.compress(self._messages)
        self._messages = list(compressed)
        pressure = self._session.check_pressure(
            usage_tokens=self._estimate_tokens()
        )
        return StreamEvent(
            action="compressed",
            compressed=True,
            tokens_saved=report.tokens_saved,
            pressure=pressure,
        )

    @property
    def messages(self) -> list[Any]:
        """Current message state (may include compressed messages)."""
        return list(self._messages)

    @property
    def facts(self) -> list[AnchorFact]:
        """Accumulated anchor facts."""
        return self._session.facts

    @property
    def session(self) -> MemoSiftSession:
        """The underlying session (for advanced configuration)."""
        return self._session

    @property
    def message_count(self) -> int:
        """Number of messages in the buffer."""
        return len(self._messages)

    def _estimate_tokens(self) -> int:
        """Estimate total token count of buffered messages."""
        contents: list[str] = []
        for msg in self._messages:
            if hasattr(msg, "content"):
                contents.append(str(msg.content))
            elif isinstance(msg, dict):
                c = msg.get("content", "")
                contents.append(str(c) if not isinstance(c, list) else str(c))
            else:
                contents.append(str(msg))
        return estimate_tokens_heuristic(contents)
